### PR TITLE
Added new DSHOT debug modes

### PR DESCRIPTION
**Pull request by damosvil**
This pr matches functionality in https://github.com/betaflight/betaflight/pull/12170 and https://github.com/betaflight/betaflight/pull/11694

Added new DSHOT EDT modes to blackbox tab. As their names are longer than usual I made the drop down list controls wider.
Also added FAILSAFE mode I found in blackbox explorer.

It can be used with Bluejay 0.20.1 RC2 or above.

<hr>
Depends on https://github.com/betaflight/betaflight/pull/12170

### CODE REVIEW

**Comment by McGiverGim** [comment]
Something is wrong, you are modifying a lot of files, including different languages files.

**Comment by damosvil** [comment]
@McGiverGim This pr adds:
- Extended dshot telemetry (EDT) checkbox enable in motors tab, so translation files need to be updated
- New debug modes to include
- ESC temperature, voltage, current, demag event, desync event, stall event, demag metric max, demag metric current, and other ESC debugging signals
- Sensors/debug functionality with new debug modes

**Comment by damosvil** [comment]
@McGiverGim I am new at adding new features to Configurator, so I may be breaking some rules I don't know. Please, could you let me know why is this wrong?

**Comment by McGiverGim** [comment]
You don't must change the locales files, only the English one. The translations are made at Crowdin and will update the others languages automatically.

**Comment by damosvil** [comment]
@McGiverGim Ok, I will restore them

File: src/js/utils/generate_filename.js
Comment (chmelevskij) [inline src/js/utils/generate_filename.js:3]: ❤️

Comment (McGiverGim) [reply src/js/utils/generate_filename.js:3]: Usually the bump version is done by ours, but this is a draft so I think is needed to see how the code looks. When final version of Configurator is released, we will bump the version and a rebase will fix the issue.

Comment (chmelevskij) [reply src/js/utils/generate_filename.js:3]: I just like ppl using modules 🤓

File: src/js/tabs/motors.js
Comment (damosvil) [reply src/js/tabs/motors.js:281]: Yes, feature 4 corresponds to previous feature 0, and feature 23 corresponds to old feature 20

Comment (haslinghuis) [reply src/js/tabs/motors.js:281]: @McGiverGim should I make a separate bugfix for this?

Comment (McGiverGim) [reply src/js/tabs/motors.js:281]: With all the changes done in firmware/configurator I'm not too sure how affects this, but if it is wrong I suppose than yes.

Comment (haslinghuis) [reply src/js/tabs/motors.js:281]: This change should not be made looking at the table:

| feature | bit | group | name |
| --: | --: | :-- | :-- |
| 0 | 0 | rxMode | RX_PPM |
| 1 | 2 | other | INFLIGHT_ACC_CALL |
| 2 | 3 | rxMode | RX_SERIAL |
| 3 | 4 | escMotorStop | MOTOR_STOP |
| 4 | 5 | other | SERVO_TILT |
| 5 | 6 | other | SOFTSERIAL |
| 6 | 7 | gps | GPS |
| 7 | 9 | other | SONAR |
| 8 | 10 | telemetry | TELEMETRY |
| 9 | 12 | 3D | 3D |
| 10 | 13 | rxMode | RX_PARALLEL_PWM |
| 11 | 14 | rxMode | RX_MSP |
| 12 | 15 | rssi | RSSI_ADC |
| 13 | 16 | other | LED_STRIP |
| 14 | 17 | other | DISPLAY |
| 15 | 18 | other | OSD |
| 16 | 20 | other | CHANNEL_FORWARDING |
| 17 | 21 | other | TRANSPONDER |
| 18 | 22 | other | AIRMODE |
| 19 | 25 | rxMode | RX_SPI |
| 20 | 27 | escSensor | ESC_SENSOR |
| 21 | 28 | antiGravity | ANTI_GRAVITY |

https://github.com/betaflight/betaflight/blob/b1433123c56598ce8ea14357d05adc9e6e1c28ca/src/main/config/feature.h#L32-L57

typedef enum {
    FEATURE_RX_PPM = 1 << 0,
    FEATURE_INFLIGHT_ACC_CAL = 1 << 2,
    FEATURE_RX_SERIAL = 1 << 3,
    FEATURE_MOTOR_STOP = 1 << 4,
    FEATURE_SERVO_TILT = 1 << 5,
    FEATURE_SOFTSERIAL = 1 << 6,
    FEATURE_GPS = 1 << 7,
    FEATURE_RANGEFINDER = 1 << 9,
    FEATURE_TELEMETRY = 1 << 10,
    FEATURE_3D = 1 << 12,
    FEATURE_RX_PARALLEL_PWM = 1 << 13,
    FEATURE_RX_MSP = 1 << 14,
    FEATURE_RSSI_ADC = 1 << 15,
    FEATURE_LED_STRIP = 1 << 16,
    FEATURE_DASHBOARD = 1 << 17,
    FEATURE_OSD = 1 << 18,
    FEATURE_CHANNEL_FORWARDING = 1 << 20,
    FEATURE_TRANSPONDER = 1 << 21,
    FEATURE_AIRMODE = 1 << 22,
    FEATURE_RX_SPI = 1 << 25,
    //FEATURE_SOFTSPI = 1 << 26, (removed)
    FEATURE_ESC_SENSOR = 1 << 27,
    FEATURE_ANTI_GRAVITY = 1 << 28,
    //FEATURE_DYNAMIC_FILTER = 1 << 29, (removed)
} features_e;

**Comment by damosvil** [comment]
@McGiverGim @haslinghuis rebased to master, ready for review

**Comment by haslinghuis** [comment]
@damosvil please fix the test

File: src/js/tabs/motors.js
Comment (haslinghuis) [inline src/js/tabs/motors.js:273]: This should be checked - as I checked the code very recent to verify assignment of featureX and I'm sure feature3 is MOTOR_STOP ===> feature = features.bit - 1

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

File: src/js/tabs/motors.js
Comment (haslinghuis) [reply src/js/tabs/motors.js:273]: Hmm, it should be 4 indeed. Having trouble with 27 getting enabled.

Comment (haslinghuis) [reply src/js/tabs/motors.js:273]: Duh, needs ESC configured on ports tab.

You were correct. See https://github.com/betaflight/betaflight-configurator/pull/3382

Comment (haslinghuis) [reply src/js/tabs/motors.js:273]: https://github.com/betaflight/betaflight-configurator/pull/3382

File: src/js/VirtualFC.js
Comment (haslinghuis) [inline src/js/VirtualFC.js:41]:
```suggestion
            use_dshot_edt: true,
            use_esc_sensor: false,
```

File: src/js/tabs/motors.js
Comment (haslinghuis) [reply src/js/tabs/motors.js:275]: :thinking: This should be 27

File: src/js/VirtualFC.js
Comment (haslinghuis) [reply src/js/VirtualFC.js:41]: It's all or nothing :)

Comment (damosvil) [reply src/js/VirtualFC.js]: done

File: src/js/tabs/motors.js
Comment (damosvil) [reply src/js/tabs/motors.js]: updated according to https://github.com/betaflight/betaflight-configurator/pull/3382

File: src/js/utils/generate_filename.js
Comment (damosvil) [reply src/js/utils/generate_filename.js]: I rolled back this change

**Comment by damosvil** [comment]
Today I generated a log in a Meteor 65 and it seems the pr works fine:
[STATUS_N_RPM_BTFL_BLACKBOX_LOG_Meteor65_20230318_104401.zip](https://github.com/betaflight/blackbox-log-viewer/files/11008198/STATUS_N_RPM_BTFL_BLACKBOX_LOG_Meteor65_20230318_104401.zip)

File: src/js/fc.js
Comment (haslinghuis) [inline src/js/fc.js:2]: API_VERSION_1_46 is not being used here

File: src/js/tabs/configuration.js
Comment (haslinghuis) [inline src/js/tabs/configuration.js:10]: API_VERSION_1_46 is not being used here

File: src/js/tabs/onboard_logging.js
Comment (haslinghuis) [inline src/js/tabs/onboard_logging.js:9]: API_VERSION_1_46 is not being used here

_We should add the values dependent on firmware version, but it has not been done before so this would need to be addressed in a separate PR._

Comment (damosvil) [reply src/js/tabs/onboard_logging.js]: fixed

File: src/js/tabs/configuration.js
Comment (damosvil) [reply src/js/tabs/configuration.js]: fixed

File: src/js/fc.js
Comment (damosvil) [reply src/js/fc.js]: fixed

**Review by haslinghuis — APPROVED** [review]

File: src/js/tabs/onboard_logging.js
Comment (haslinghuis) [inline src/js/tabs/onboard_logging.js:343]: These 3 are in master now.

Comment (nerdCopter) [reply src/js/tabs/onboard_logging.js:343]: @damosvil , merge master into branch or rebase branch upon master should fix it.  i recall betaflight team prefers rebase, if i'm not mistaken.

Comment (damosvil) [reply src/js/tabs/onboard_logging.js]: rebased to current master

File: src/js/tabs/motors.js
Comment (haslinghuis) [inline src/js/tabs/motors.js:690]: 690, 697 and 700 should have it's own `if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_46) {` block

Comment (damosvil) [reply src/js/tabs/motors.js]: done

File: src/js/tabs/motors.js
Comment (damosvil) [reply src/js/tabs/motors.js]: done

File: src/js/tabs/onboard_logging.js
Comment (haslinghuis) [inline src/js/tabs/onboard_logging.js:332]:
```suggestion
                {text: "RPM_LIMIT"},
```

Comment (damosvil) [reply src/js/tabs/onboard_logging.js]: Fixed

**Review by haslinghuis — APPROVED** [review]

**Comment by sugaarK** [comment]
we still getting this done? rc1 is planned for Dec 1

**Comment by damosvil** [comment]
The objective of this branch is allowing to log esc events in the blackbox to know how well commutation goes in the escs.
I can work in the rebase, but at this time Bluejay is the only fw that fully implements extended telemetry and it seems that no other firmwares are interested in implementing it, neither any company, so if you like I could keep a branch in my own repo for BF4.5 and then we can let it go.
Probably this is not the time for this class of innovation.

**Comment by sugaarK** [comment]
> The objective of this branch is allowing to log esc events in the blackbox to know how well commutation goes in the escs.
> I can work in the rebase, but at this time Bluejay is the only fw that fully implements extended telemetry and it seems that no other firmwares are interested in implementing it, neither any company, so if you like I could keep a branch in my own repo for BF4.5 and then we can let it go.
> Probably this is not the time for this class of innovation.

If it’s ready passes code review and you will make use of it then I say let’s go.. I’m sure I can convince am32 to get on it and venders too

File: src/js/tabs/motors.js
Comment (haslinghuis) [inline src/js/tabs/motors.js:820]: Maybe just

```suggestion
            $('.checkboxDshotEdt').toggle(protocolConfigured && digitalProtocol && dshotBidirElement.is(':checked') && !$("input[name='ESC_SENSOR']").is(':checked') && semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_46));
```

Comment (damosvil) [reply src/js/tabs/motors.js]: refactored

**Comment by ctzsnooze** [comment]
Thanks for keeping active here.  I look forward to testing this with a suitable bluejay version.

**Review by haslinghuis — APPROVED** [review]

**Review by McGiverGim — APPROVED** [review]

File: src/js/tabs/motors.js
Comment (chmelevskij) [inline src/js/tabs/motors.js:764]: Not sure about the code style here.

Better to use `this` or similar way to get the info about the actual order, sometimes things get weird with closures in js.

Also, not sure what triggering `change` event and then adding `on` handler does here 🤔

Something like the following is a bit more streamlined/usual in js land
```suggestion
const dshotEditCheckbox = $('#dshotEdt');

dshotEditCheckbox.prop('checked', FC.MOTOR_CONFIG.use_dshot_edt)
    .on("change", function () {
        const isChecked = $(this).is(':checked');
        const newValue = isChecked !== FC.MOTOR_CONFIG.use_dshot_edt ? 'On' : 'Off';

        self.analyticsChanges['DshotEdt'] = newValue;
        FC.MOTOR_CONFIG.use_dshot_edt = isChecked;
    })
    .trigger("change");
```

File: src/tabs/motors.html
Comment (chmelevskij) [inline src/tabs/motors.html:96]: Ooof, these floats look painful 🫠 But probably better for style consistency. More of a note to self/team 🙈

**Review by chmelevskij — APPROVED** [review]

**Comment by nerdCopter** [comment]
time to rebase and consider review comments, TYVM

**Comment by haslinghuis** [comment]
Closing as firmware PR got stalled.

**Comment by nerdCopter** [comment]
@damosvil , please reopen if this is still a desirable PR.  after re-opening please merge master or rebase master.

**Comment by nerdCopter** [comment]
reopened, as i plan update the PR/branch.

**Review by nerdCopter — CHANGES_REQUESTED** [review]

File: src/js/tabs/motors.js
Comment (haslinghuis) [reply src/js/tabs/motors.js:15]: `API_VERSION_1_47` needs to be added instead.

**Comment by haslinghuis** [comment]
This work will be continued in a new PR.

**Pull request closed without merging**
